internal/generate/olm-catalog/descriptor/crd.go

```diff
@@ -0,0 +1,210 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/fatih/structtag"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	"k8s.io/apimachinery/pkg/version"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"github.com/operator-framework/operator-sdk/internal/util/k8sutil"
+)
+
+// CRDDescriptionToKey creates a key unique to a crdDescription.
+func CRDDescriptionToKey(crd operatorsv1alpha1.CRDDescription) registry.DefinitionKey {
+	return registry.DefinitionKey{
+		Name:    crd.Name,
+		Group:   NameToFullGroup(crd.Name),
+		Version: crd.Version,
+		Kind:    crd.Kind,
+	}
+}
+
+// KeyToCRDDescription creates a crdDescription from a unique key.
+func KeyToCRDDescription(key registry.DefinitionKey) operatorsv1alpha1.CRDDescription {
+	return operatorsv1alpha1.CRDDescription{
+		Name:        key.Name,
+		Version:     key.Version,
+		Kind:        key.Kind,
+		DisplayName: k8sutil.GetDisplayName(key.Kind),
+	}
+}
+
+// NameToFullGroup returns everything but the first element of a CRD name,
+// which by definition is >resource>.<full group>.
+func NameToFullGroup(name string) string {
+	return splitFirst(name, ".", 1)
+}
+
+// FullGroupToGroup returns the first element of an API group, ex. "foo" of
+// "foo.example.com".
+func FullGroupToGroup(group string) string {
+	return splitFirst(group, ".", 0)
+}
+
+// splitFirst splits s by the first sep encountered and returns the first
+// (half = 0) or second (half = 1) element of the result.
+func splitFirst(s, sep string, half int) string {
+	if split := strings.SplitN(s, sep, 2); len(split) == 2 {
+		return split[half]
+	}
+	return s
+}
+
+// buildCRDDescriptionFromType builds a crdDescription for the Go API defined
+// by key from markers and type information in g.types.
+func (g generator) buildCRDDescriptionFromType(key registry.DefinitionKey,
+	kindType *markers.TypeInfo) (operatorsv1alpha1.CRDDescription, error) {
+
+	// Initialize the description.
+	description := KeyToCRDDescription(key)
+	description.Description = kindType.Doc
+
+	// Parse resources and displayName from the kind type's markers.
+	for _, markers := range kindType.Markers {
+		for _, marker := range markers {
+			d := marker.(Description)
+			switch {
+			case d.DisplayName != "":
+				description.DisplayName = d.DisplayName
+			case len(d.Resources) != 0:
```

Comment:
I am not sure if we should switch here. Cannot both scenarios be true?

Reply:
Yep, I forgot to update this from legacy-style code. Nice catch.